ui/desktop/src/components/settings/tunnel/TunnelSection.tsx

```diff
@@ -0,0 +1,384 @@
+  const [showQRModal, setShowQRModal] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+  const [copiedUrl, setCopiedUrl] = useState(false);
+  const [copiedSecret, setCopiedSecret] = useState(false);
+  const [showDetails, setShowDetails] = useState(false);
+  const [tunnelMode, setTunnelMode] = useState<TunnelMode>('lapstone');
+  const [showOtherOptions, setShowOtherOptions] = useState(false);
+
+  useEffect(() => {
+    const loadTunnelMode = async () => {
+      try {
+        const mode = await window.electron.getTunnelMode();
+        setTunnelMode(mode as TunnelMode);
+      } catch (err) {
+        console.error('Error loading tunnel mode:', err);
+      }
+    };
+    loadTunnelMode();
+  }, []);
+
+  useEffect(() => {
+    checkTunnelStatus();
+  }, []);
+
+  const checkTunnelStatus = async () => {
+    try {
+      const status = await window.electron.getTunnelStatus();
+      setTunnelStatus(status as TunnelStatus);
+    } catch (err) {
+      console.error('Error checking tunnel status:', err);
+    }
+  };
+
+  const handleStartTunnel = async () => {
+    setError(null);
+    setTunnelStatus({ state: 'starting', info: null });
+
+    try {
+      const tunnelInfo = await window.electron.startTunnel();
+      setTunnelStatus({ state: 'running', info: tunnelInfo as TunnelInfo });
+      setShowQRModal(true);
+    } catch (err) {
+      console.error('Error starting tunnel:', err);
+      setError(err instanceof Error ? err.message : 'Failed to start tunnel');
+      setTunnelStatus({ state: 'error', info: null });
+    }
+  };
+
+  const handleStopTunnel = async () => {
+    try {
+      await window.electron.stopTunnel();
+      setTunnelStatus({ state: 'idle', info: null });
+      setShowQRModal(false);
+    } catch (err) {
+      console.error('Error stopping tunnel:', err);
+      setError(err instanceof Error ? err.message : 'Failed to stop tunnel');
+    }
+  };
+
+  const handleTunnelModeChange = async (mode: TunnelMode) => {
+    try {
+      await window.electron.setTunnelMode(mode);
+      setTunnelMode(mode);
+    } catch (err) {
+      console.error('Error setting tunnel mode:', err);
+    }
+  };
+
+  const copyToClipboard = async (text: string, type: 'url' | 'secret') => {
+    try {
+      await navigator.clipboard.writeText(text);
+      if (type === 'url') {
+        setCopiedUrl(true);
+        setTimeout(() => setCopiedUrl(false), 2000);
+      } else {
+        setCopiedSecret(true);
+        setTimeout(() => setCopiedSecret(false), 2000);
+      }
+    } catch (err) {
+      console.error('Failed to copy to clipboard:', err);
+    }
+  };
+
+  // Generate QR code data
+  const getQRCodeData = () => {
+    if (!tunnelStatus.info) return '';
+
+    // Use the actual tunnel URL (works for both Tailscale and Lapstone)
+    // For Tailscale: uses ipv4 if available, otherwise falls back to url
+    // For Lapstone: always uses the public Cloudflare URL
```

Comment:
this comment is harder to read than the actual code, so I would drop it. Since we're getting this from the other side anyway, why not always have a url in tunnelStatus and we can worry about it there?

Reply:
heh - this one was actually, ironically by me (and left for me!) yes is confusing.